src/frontend/src/optimizer/property/func_dep.rs

```diff
@@ -0,0 +1,373 @@
+                tmp.set(i, true);
+            }
+            tmp
+        };
+        FunctionalDependency { from, to }
+    }
+
+    /// Create a [`FunctionalDependency`] with a key. The combination of these columns can determine
+    /// all other columns.
+    fn with_key(column_cnt: usize, key_indices: &[usize]) -> Self {
+        let mut from = FixedBitSet::with_capacity(column_cnt);
+        for &idx in key_indices {
+            from.set(idx, true);
+        }
+        let mut to = from.clone();
+        to.toggle_range(0..to.len());
+        FunctionalDependency { from, to }
+    }
+
+    /// Create a [`FunctionalDependency`] for constant columns.
+    /// These columns can be determined by any column.
+    pub fn with_constant(column_cnt: usize, constant_indices: &[usize]) -> Self {
+        let mut to = FixedBitSet::with_capacity(column_cnt);
+        for &i in constant_indices {
+            to.set(i, true);
+        }
+        FunctionalDependency {
+            from: FixedBitSet::with_capacity(column_cnt),
+            to,
+        }
+    }
+
+    pub fn into_parts(self) -> (FixedBitSet, FixedBitSet) {
+        (self.from, self.to)
+    }
+}
+
+/// [`FunctionalDependencySet`] contains the functional dependencies.
+///
+/// It is used in optimizer to track the dependencies between columns.
+#[derive(Debug, PartialEq, Clone, Default)]
+pub struct FunctionalDependencySet {
+    /// `strict` contains all strict functional dependencies.
+    ///
+    /// The strict functional dependency use the **NULL=** semantic. It means that all NULLs are
+    /// considered as equal. So for following table, A --> B is not valid.
+    /// **NOT** allowed.
+    /// ```text
+    ///   A   | B
+    /// ------|---
+    ///  NULL | 1
+    ///  NULL | 2
+    /// ```
+    ///
+    /// Currently we only have strict dependencies, but we may also have lax dependencies in the
+    /// future.
+    strict: Vec<FunctionalDependency>,
+}
+
+impl fmt::Display for FunctionalDependencySet {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("{")?;
+        for fd in self.strict.iter().with_position() {
+            match fd {
+                itertools::Position::First(fd) | itertools::Position::Middle(fd) => {
+                    f.write_fmt(format_args!("{}, ", fd))?;
+                }
+                itertools::Position::Last(fd) | itertools::Position::Only(fd) => {
+                    f.write_fmt(format_args!("{}", fd))?;
+                }
+            }
+        }
+        f.write_str("}")
+    }
+}
+
+impl FunctionalDependencySet {
+    /// Create a empty [`FunctionalDependencySet`]
+    pub fn new() -> Self {
+        Self { strict: Vec::new() }
+    }
+
+    /// Create a [`FunctionalDependencySet`] with the indices of a key.
+    ///
+    /// The **combination** of these columns can determine all other columns.
+    ///
+    /// Note that if `key_indices` empty, an empty fd set will be returned.
```

Comment:
When will there be an empty key?

Reply:
In some cases, we use `pk_indices` to initialize a func_dep_set such as `logical_scan`. And the `pk_indices` could be empty. However, an empty `pk_indices` means that "pk cannot be derived" rather than "all columns of this relation are constant".

Reply:
If so, would it be better to add branching logic when pk cannot be derived, and assert key is non empty here?